serde_derive/src/bound.rs

```diff
@@ -50,18 +50,35 @@ pub fn with_where_predicates(
     generics
 }
 
-pub fn with_where_predicates_from_fields<F>(
+pub fn with_where_predicates_from_fields<F, W>(
     cont: &Container,
     generics: &syn::Generics,
+    trait_bound: &syn::Path,
     from_field: F,
+    gen_bound_where: W,
 ) -> syn::Generics
 where
     F: Fn(&attr::Field) -> Option<&[syn::WherePredicate]>,
+    W: for<'a> Fn(&attr::Field) -> bool,
 {
     let predicates = cont.data
         .all_fields()
-        .flat_map(|field| from_field(&field.attrs))
-        .flat_map(|predicates| predicates.to_vec());
+        .flat_map(|field| {
+            let field_ty = field.ty;
+            let field_bound: Option<syn::WherePredicate> = match (field_ty, gen_bound_where(&field.attrs)) {
+                (&syn::Type::Path(ref ty_path), true) => match (ty_path.path.segments.first(), generics.params.first()) {
```

Comment:
Please make this work even if the associated type is not from the first generic parameter.

```rust
#[derive(Serialize, Deserialize)]
struct AssocDerive<S, T: AssocSerde> {
    s: S,
    assoc: T::Assoc,
}
```